### ISSUE TITLE
1.24.3 leaks pidns with cgroup v2

### ISSUE THREAD

**Issue by rkojedzinszky** [cri-o/cri-o#6349]
### What happened?

Observed that on a node where many cronjobs get scheduled (i.e.many pods being created than stopped), syslog entries appear:
```
Nov 07 08:22:18 k8s-node16 kubelet[699]: W1107 08:22:18.605965     699 container.go:589] Failed to update stats for container "/pids/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podeb779fdd_e230_4fb0_b027_f9f91f117b46.slice/crio-f1800404321299f38a4ba5dfedf95df94b825507597527b48bf3a6d9f2f3eb13.scope": unable to determine device info for dir: /var/lib/containers/storage/overlay/304bfd8df6d35aaf9a5632b9bcdc69bdd54aa9277baf2c18c4930072d017514c/diff: stat failed on /var/lib/containers/storage/overlay/304bfd8df6d35aaf9a5632b9bcdc69bdd54aa9277baf2c18c4930072d017514c/diff with error: no such file or directory, continuing to push stats
```

That is because entries under `/sys/fs/cgroup/pids/kubepods.slice/kubepods-burstable.slice/` are not cleaned up. I suspect the count of `crictl pods` and entries under the mentioned dir should be in sync, but I may be wrong.

```
# crictl pods
POD ID              CREATED             STATE               NAME                     NAMESPACE           ATTEMPT             RUNTIME
6b309d1ef5223       13 minutes ago      Ready               kube-router-dr2vh        kube-system         0                   (default)
41b1338d583b2       13 minutes ago      Ready               truenas-csi-node-5lkh5   truenas-csi         0                   (default)
b12fd8fb24ccb       13 minutes ago      Ready               promtail-vfflj           monitoring          0                   (default)
950e109f187a0       13 minutes ago      Ready               node-exporter-km99c      kube-system         0                   (default)
# ls -la /sys/fs/cgroup/pids/kubepods.slice/kubepods-burstable.slice/
total 0
drwxr-xr-x 11 root root 0 Nov  7 08:13 .
drwxr-xr-x  3 root root 0 Nov  7 08:13 ..
-r--r--r--  1 root root 0 Nov  7 08:13 cgroup.controllers
-r--r--r--  1 root root 0 Nov  7 08:13 cgroup.events
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.freeze
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.max.depth
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.max.descendants
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.procs
-r--r--r--  1 root root 0 Nov  7 08:13 cgroup.stat
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.subtree_control
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.threads
-rw-r--r--  1 root root 0 Nov  7 08:13 cgroup.type
-rw-r--r--  1 root root 0 Nov  7 08:13 cpu.pressure
-r--r--r--  1 root root 0 Nov  7 08:13 cpu.stat
-rw-r--r--  1 root root 0 Nov  7 08:13 io.pressure
drwxr-xr-x  3 root root 0 Nov  7 08:34 kubepods-burstable-pod321a0667_9839_4537_8dfe_8a2f42ed2a01.slice
drwxr-xr-x  3 root root 0 Nov  7 08:13 kubepods-burstable-pod571e43da_a130_438b_86d8_5b5fea8c2180.slice
drwxr-xr-x  3 root root 0 Nov  7 08:13 kubepods-burstable-pod57769bbc_9867_42a7_83bd_7a65fc4791ab.slice
drwxr-xr-x  3 root root 0 Nov  7 08:29 kubepods-burstable-pod5872eef3_b96b_488c_8338_265e32a9b87c.slice
drwxr-xr-x  3 root root 0 Nov  7 08:16 kubepods-burstable-podb2f5e0d7_a53b_4a33_a4ab_ca4ac994e365.slice
drwxr-xr-x  3 root root 0 Nov  7 08:13 kubepods-burstable-podb3174e7f_0cb1_47d5_8b6e_5c1b9904f07e.slice
drwxr-xr-x  3 root root 0 Nov  7 08:34 kubepods-burstable-podb3e4520f_03ad_46e0_bfcc_c1dc739260e8.slice
drwxr-xr-x  3 root root 0 Nov  7 08:16 kubepods-burstable-podeb779fdd_e230_4fb0_b027_f9f91f117b46.slice
drwxr-xr-x  3 root root 0 Nov  7 08:13 kubepods-burstable-podec55611f_6ee5_481f_9e8b_f6ae883baf29.slice
-rw-r--r--  1 root root 0 Nov  7 08:13 memory.pressure
```

Howewer I have 4 running pods, and 9 entries under /sys. kubelet/cadvisor then logs warning messages for pid namespaces not cleaned up, and this floods my logs.

### What did you expect to happen?

Expected every resource cleaned up after pod termination, logs being not flooded.

### How can we reproduce it (as minimally and precisely as possible)?

Have cgroups v2 with crio-1.24. Start and stop a pod on the node, and observe entries not being removed from `/sys/fs/cgroup/pids/kubepods.slice/kubepods-burstable.slice`.

### Anything else we need to know?

It is to be noted that with v1.25 `/sys/fs/cgroup/pids` just does not exist.

### CRI-O and Kubernetes version

<details>

```console
# crio --version
crio version 1.24.3
Version:          1.24.3
GitCommit:        251db4e7a44dd5b76586bd2f8d818691ed8413f7
GitTreeState:     clean
BuildDate:        2022-11-07T06:13:20Z
GoVersion:        go1.18.8
Compiler:         gc
Platform:         linux/amd64
Linkmode:         dynamic
BuildTags:        apparmor, exclude_graphdriver_devicemapper, containers_image_openpgp, containers_image_ostree_stub, seccomp, selinux
SeccompEnabled:   true
AppArmorEnabled:  true
```

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:57:26Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:49:09Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/arm64"}
```
</details>

### OS version

<details>

```console
# On Linux:
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
# uname -a
Linux k8s-node16 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
```

</details>

### Additional environment details (AWS, VirtualBox, physical, etc.)

<details>
Bare metal
</details>

**Comment by anthr76** [comment]
I'm also facing this issue.

**Comment by Quarky9** [comment]
Can confirm the exact same issue on 20+ clusters running 1.23 as well as 1.24.3 .

We implemented a daily cleanup job to remove the leaked cgroup folders to prevent the kubelet failing eventually.

**Comment by dmamchyts** [comment]
any updates?

**Comment by n4j** [comment]
I ran cri-o `1.24` from source and created/deletd a pod with `crictl` in my test bed.

The cgroups were getting cleaned-up as soon as I stopped the pod + I didn't see cgroup path like this `/sys/fs/cgroup/pids/`.

Let me try to setup k8s in my test bed and replicate the issue.

I'll keep this thread posted of my observations

**Comment by n4j** [comment]
I am able to reproduce the issue in my test bed.

I have installed kubernetes (v1.26.1) via kubeadm  and cri-o from the source code (branch is release-1.24.0).

And the issue is reproduced  every time I create & delete the container.

I am planning to put some debug logs in kubelet/cri-o to better understand what's going on here.

**Comment by rphillips** [comment]
How were the cronjobs created? Were they created with `kubectl create job`? Creating the cronjob in this manner does not create the job within the controller for cleanup.

**Comment by n4j** [comment]
I was able to figure the root cause of this issue

TL;DR - The issue stems from the fact that cgroup path created is different from what `runc` is attempting to delete.

I still have to figure which component is responsible for the mismatch but here's my analysis so far

In my setup, I had created a busybox container with following PodSpec

```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: sleepy-v11
spec:
  containers:
    - name: test
      image: almalinux:8
      command: ["/bin/sleep", "10000"]
```



The k8s UID of the pod was `546e0e73-40dd-42e4-b6ce-0b18b5cc49f7`  and querying `crictl` yielded container id `3d689b62983aa6d734b72ed1e368c8f7a5c4ba1eb10a9bf878535d9d7f2922f1` and podSandboId `97121559b173d96097a8fec88ef9c26b8040d82985735ca49836ba963bed60a6` 

So far, we have below IDs queried from `kubectl` & `crictl`

```bash
K8S_POD_UID=546e0e73-40dd-42e4-b6ce-0b18b5cc49f7
CRIO_CONTAINER_ID=3d689b62983aa6d734b72ed1e368c8f7a5c4ba1eb10a9bf878535d9d7f2922f1
CRIO_POD_SANDBOX_ID=97121559b173d96097a8fec88ef9c26b8040d82985735ca49836ba963bed60a6
```

The below `cgroupv2` hierarchy was created under `/sys/fs/cgroup/pids`

```
kubepods-besteffort-pod546e0e73_40dd_42e4_b6ce_0b18b5cc49f7.slice
├── cgroup.controllers
├── cgroup.events
├── cgroup.freeze
├── cgroup.kill
├── cgroup.max.depth
├── cgroup.max.descendants
├── cgroup.procs
├── cgroup.stat
├── cgroup.subtree_control
├── cgroup.threads
├── cgroup.type
├── cpu.pressure
├── cpu.stat
├── crio-97121559b173d96097a8fec88ef9c26b8040d82985735ca49836ba963bed60a6.scope
│             ├── cgroup.controllers
│             ├── cgroup.events
│             ├── cgroup.freeze
│             ├── cgroup.kill
│             ├── cgroup.max.depth
│             ├── cgroup.max.descendants
│             ├── cgroup.procs
│             ├── cgroup.stat
│             ├── cgroup.subtree_control
│             ├── cgroup.threads
│             ├── cgroup.type
│             ├── cpu.pressure
│             ├── cpu.stat
│             ├── io.pressure
│             └── memory.pressure
├── io.pressure
└── memory.pressure
```
Note that the format of the cgroup hierarchy is of format `/sys/fs/cgroup/pids/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod${K8S_POD_UID}.slice/crio-${CRIO_POD_SANDBOX_ID}.scope`

I observed that when request is received by `runc` it infers the cgroup path of the pod to be `/sys/fs/cgroup/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod546e0e73_40dd_42e4_b6ce_0b18b5cc49f7.slice/crio-3d689b62983aa6d734b72ed1e368c8f7a5c4ba1eb10a9bf878535d9d7f2922f1.scope`

The format of the generated path is `/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod${K8S_POD_UID}.slice/crio-${CRIO_CONTAINER_ID}.scope`

Since this is not the actual path of the `cgroup`, ` runc`'s  [rmdir](https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L231) method fails with error `sys.ERRNOENT` and this err is swallowed in the code and hence it's not returned to the caller

An extra bit of information, the same `runc` infers the same cgroup path while creating and deleting containers + when runc is executed to delete the container, it's passed `CRIO_CONTAINER_ID` as an argument.

**Comment by n4j** [comment]
Ok, so I believe here is what's causing the issue

 In `/server/sandbox_run_linux.go#runPodSandbox`  we have below at [line 907](https://github.com/cri-o/cri-o/blob/release-1.24/server/sandbox_run_linux.go#L907)

```golang
if err := s.config.CgroupManager().CreateSandboxCgroup(cgroupParent, sbox.ID()); err != nil {
			return nil, errors.Wrapf(err, "create dropped infra %s cgroup", sbox.ID())
		}
```
In call to `CreateSandboxCgroup` we are passing `sbox.ID()` which is `podSandboxId` while 

while in `/internal/config/cgmgr/systemd.go` [line 203 ](https://github.com/cri-o/cri-o/blob/release-1.24/internal/config/cgmgr/systemd.go#L203 ) expects a `containerId` as second argument

```golang
func (m *SystemdManager) CreateSandboxCgroup(sbParent, containerID string) error {
	return createSandboxCgroup(sbParent, containerID, m)
}
```

Therefore, the path is getting created with podSandboxId and when `runc` tries to delete it, it looks for the one with `containerId` and hence the leak.

@haircommander What's your opinion on this? If the above RCA looks good to you then I can raise a PR for it.

**Comment by haircommander** [comment]
so in this case, sbox.ID() is the ID of the infra container (which we're spoofing in this case). Though, it's possible we never call delete on sbox.ID(). We may need to add that on either stop or remove pod sandbox call

**Comment by n4j** [comment]
When the Pod Sandbox is deleted via [RemovePodSandbox](https://github.com/cri-o/cri-o/blob/release-1.24/server/sandbox_remove.go#L17), cgroups for PodSandbox aren't cleaned up and hence the leak.

I'll submit a PR for PodSandbox cgroup cleanup.

**Comment by n4j** [comment]
The issue has been fixed and merged to branches `main`, `release-1.25`, `release-1.24` and `release-1.23`
